content/blog/2023-12-11-rvv-llvm-gisel/index.md

```diff
@@ -0,0 +1,177 @@
++++
+title = "Support LLVM GlobalISel for RISC-V Vector Extension on Partial ALU Instructions"
+[extra]
+bio = """
+  Jiahan Xie is a first-year CS MS student at Cornell whose interests lie in Compilers and Computer Architecture.
+
+  Michael Maitland was a student in Cornell CS 6120 Advanced Compilers and he's now working at SiFive.
+"""
+[[extra.authors]]
+name = "Jiahan Xie"
+[[extra.authors]]
+name = "Michael Maitland"
++++
+
+# Introduction
+
+The open [RISC-V instruction set architecture (ISA)](https://riscv.org/technical/specifications/) has an interesting extension, [the RISC-V "V" Vector Extension](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc). The unique part of RISC-V Vector Extension is that its vector registers have flexible widths, VLEN, which makes programming in RISC-V Vector Extension agnostic to the vector register sizes. This feature really distinguishes the RISC-V Vector Extension from the traditional SIMD extensions, such as [x86 SSE](https://en.wikipedia.org/wiki/Streaming_SIMD_Extensions) (with a fixed size 128-bit vector length)/[AVX](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions) (256-bit), and [Arm NEON](https://developer.arm.com/Architectures/Neon) (128-bit). The increasing vector lengths can pose a challenge to the traditional SIMD extensions as they have to address compatability and support all existing fixed size vector lengths in their ISAs. On the contrary, with the vector lengths agnostic principles, binary code generated by RISC-V assembly is automatically portable between different CPUs.
```

Comment:
RISC-V folks more commonly refer to VL as the unique part.

> The vl register holds an unsigned integer specifying the number of elements to be updated with results from a vector instruction

This is in contrast to many SIMD approaches where the instruction pneumonic and different sizes of register impose a fixed number of elements processed. For example, the [ARM Neon VADD](https://developer.arm.com/documentation/den0018/a/NEON-Intrinsics-Reference/Arithmetic/VADD).

On a given hardware implementation, the widths of registers in RISCV are flexible due to LMUL. This is also unique to RISCV. 

The VLEN however, is fixed for that hardware implementation.